### PR TITLE
[move] [read/write set] Use temp_indices instead during footprint substitution

### PR DESCRIPTION
**Pull request by danhper**
## Motivation

During footprint in `access_paths.rs`, using `actuals` caused nested memory access to not be resolved correctly.
We now use the callee formal indices instead and when reading from the caller locals.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?
yes

## Test Plan

Added e2e tests

## Related PRs

#8658 
/cc @sblackshear

### CODE REVIEW

**Comment by danhper** [comment]
@sblackshear I just realized there's a failing tests for the dynamic analysis part but I'm not 100% sure what the issue is.
I thought that it was because we are passing an empty mapping to `sub_actuals` here: https://github.com/diem/diem/pull/8697/files#diff-f6ed2a6610a79c36bc912f9f7ebf87ee1952d4b0ebede36ff82dea0809e7ea7eR115
but binding actuals to the passed submap did not do the trick so I might have missed something.
This is what I tried to add to `from_args_`:

```rust
let mut empty_sub_map: AccessPathTrie<AbsAddr> = AccessPathTrie::default();
for (i, actual) in actuals.iter().enumerate() {
      empty_sub_map.bind_local(i, actual.clone(), fun_env);
}
```

I'll continue to debug this but please if me know if you spot the issue.
Thanks!

**Comment by sblackshear** [comment]
> @sblackshear I just realized there's a failing test for the dynamic analysis part but I'm not 100% sure what the issue is.
> I thought that it was because we are passing an empty mapping to `sub_actuals` here: https://github.com/diem/diem/pull/8697/files#diff-f6ed2a6610a79c36bc912f9f7ebf87ee1952d4b0ebede36ff82dea0809e7ea7eR115
> but binding actuals to the passed submap did not do the trick so I might have missed something.
> This is what I tried to add to `from_args_`:
> 
> ```rust
> let mut empty_sub_map = AccessPathTrie::default();
> for (i, actual) in actuals.iter().enumerate() {
>       empty_sub_map.bind_local(i, actual.clone(), fun_env);
> }
> ```
> 
> I'll continue to debug this but please if me know if you spot the issue.
> Thanks!

I think the issue might be that you also need to add `signers` to `empty_sub_map`. If the function in question is (e.g.) `foo(a: signer, b: signer, c: address, d: u64)`, `actuals` will contain `c` and `d`, but not `a` and `b`. I think the following code (or a variant) should work:

```
let mut empty_sub_map = AccessPathTrie::default();
for (i, signer) in signers.iter().enumerate() {
  empty_sub_map.bind_local(i, signer.clone(), fun_env);
}

for (actual_index, ty) in formal_types[num_signers..].iter().enumerate() {
  let actual = ...; // existing code
  empty_sub_map.bind_local(i, actual.clone(), fun_env); // new code
}
```

File: language/move-prover/bytecode/src/read_write_set_analysis.rs [hunk 1]
```diff
@@ -83,12 +83,13 @@ impl ReadWriteSetState {
         accesses: AccessPathTrie<Access>,
         actuals: &[AbsAddr],
         type_actuals: &[Type],
+        temp_indices: &[TempIndex],
```
Comment (sblackshear) [inline language/move-prover/bytecode/src/read_write_set_analysis.rs:86]: Not sure if this is the bug, but something I noticed we shouldn't need `actuals: &[AbsAddr]` here anymore (and same with `apply_summary`)--the code can read the actual from `sub_map` using the temp indexes.

Comment (danhper) [reply language/move-prover/bytecode/src/read_write_set_analysis.rs:86]: Good point, thanks! I'll refactor that!

File: language/move-prover/bytecode/src/read_write_set_analysis.rs [hunk 2]
```diff
@@ -144,15 +146,17 @@ impl ReadWriteSetState {
         let callee_summary = callee_summary_.clone();
         // (1) bind all footprint values and types in callee locals to their caller values
         let mut new_callee_locals = callee_summary.locals.substitute_footprint(
-            actuals,
             type_actuals,
+            &temp_indices,
+            callee_fun_env,
```
Comment (sblackshear) [inline language/move-prover/bytecode/src/read_write_set_analysis.rs:151]: I think this should be `caller_fun_env`--we are replacing footprint values in the callee summary with caller state.

File: language/move-prover/bytecode/src/read_write_set_analysis.rs [hunk 3]
```diff
@@ -144,15 +146,17 @@ impl ReadWriteSetState {
...
             &self.locals,
             AbsAddr::substitute_footprint,
         );
         // (2) bind all footprint values and types in callee accesses to their caller values
         let mut new_callee_accesses = callee_summary.accesses.substitute_footprint_skip_data(
-            actuals,
             type_actuals,
+            &temp_indices,
+            callee_fun_env,
```
Comment (sblackshear) [inline language/move-prover/bytecode/src/read_write_set_analysis.rs:159]: Same here.

File: language/move-prover/bytecode/src/read_write_set_analysis.rs (hunk 2 above)
Comment (danhper) [reply language/move-prover/bytecode/src/read_write_set_analysis.rs:151]: I am not sure I understand why it should be the caller env rather than the callee's one.
The only place where this `FunctionEnv` is actually used is in `AccessPath::substitute_footprint` as an argument to `Root::from_index`, so to check whether the given index is a formal or a local.
I would think that we need the callee's env to know whether the given index is a formal or not. Or am I getting mixed up here?

File: language/move-prover/bytecode/src/access_path.rs [hunk 4]
```diff
@@ -636,36 +646,32 @@ impl AccessPath {
...
-                if !actual.is_empty() {
-                    let new_ap = AccessPath::new(self.root.clone(), new_offsets);
-                    acc.join(&new_ap.prepend_addrs(actual));
-                } else {
-                    // TODO: change actuals to &[TempIndex], look up actuals[i]/new_offsets in sub_map")
-                }
+                let temp_index = temp_indices[*i];
+                let root = Root::from_index(temp_index, func_env);
+                let ap = AccessPath::new(root, new_offsets);
+                sub_map.get_access_path(ap).cloned().unwrap_or_default()
```
Comment (sblackshear) [inline language/move-prover/bytecode/src/access_path.rs:663]: This doesn't look quite right to me. At a high level, what this code is doing is:

- `self` is a footprint path in the callee state that reads from the environment (e.g., `Formal(0)/f/g)`.
- We want to resolve the formal in the caller environment (`sub_map`), then prepend the result to the callee's path. For example, if `Formal(0)` is bound to `Formal(1)/e` in the caller, we would like this function to produce the path `Formal(1)/e/f/g`. In the old code, this is accomplished via `prepend_addrs(Formal(0)/f/g, Formal(1)/e)`. But the new code drops the call to `prepend_addrs`.

File: language/move-prover/bytecode/src/read_write_set_analysis.rs (hunk 2 above)
Comment (sblackshear) [reply language/move-prover/bytecode/src/read_write_set_analysis.rs:151]: Here is my reasoning. If we have the following code:

```
fun caller(tmp1) {
  tmp2 = @0x2
  foo(tmp1, tmp2)
}
```

`temp_indices` should be `[tmp1, tmp2]`. When we get to `substitute_footprint`, we will need to resolve the values of `tmp1` and `tmp2` from the caller state. We do this by constructing an access path from `Root::from_index(tmp1, fun_env)`, then reading the value of that access path in the caller state. We want `tmp1` to be read as a `Formal` and `tmp2` to be read as a `Local`, and we need the caller's function environment to determine this.

Comment (danhper) [reply language/move-prover/bytecode/src/read_write_set_analysis.rs:151]: I see, that makes sense, thanks!

File: language/move-prover/bytecode/src/access_path.rs (hunk 4 above)
Comment (danhper) [reply language/move-prover/bytecode/src/access_path.rs]: This was slightly nastier to fix than I hoped but I believe the latest version should cover the cases that did not work in both the previous version of this PR and the previous implementation.
At least, it does seem to fix the issue mentioned here: https://github.com/diem/diem/pull/8706/files#r666360471

File: language/move-prover/bytecode/src/access_path.rs [hunk 5]
```diff
@@ -636,36 +646,64 @@ impl AccessPath {
...
-                }
+                let temp_index = temp_indices[*i];
+                let caller_root = Root::from_index(temp_index, func_env);
+                let mut offsets = vec![];
+
+                // In this loop, we lookup the access path in the submap,
+                // starting with the root of the caller and adding the
+                // offsets of the current access path until a result is found
+                // or we do not have any more offsets to add
+                let addrs = loop {
```
Comment (sblackshear) [inline language/move-prover/bytecode/src/access_path.rs:668]: Can this be a loop over `new_offsets`?

File: language/move-prover/bytecode/src/access_path.rs [hunk 6]
```diff
@@ -659,40 +659,29 @@ impl AccessPath {
...
-                    // remove(0) is inefficient but should not be problematic given offsets is very small
-                    offsets.push(new_offsets.remove(0));
-                };
-
-                // new_offsets will either be the full offsets of this access path
-                // or a subset of them if we had to use some of them to read
-                // the access path
-                AccessPath::new(self.root.clone(), new_offsets).prepend_addrs(addrs)
+                }
+                results
```
Comment (danhper) [inline language/move-prover/bytecode/src/access_path.rs:684]: @sblackshear Hopefully, this should cover the edge case we just discussed! I'll add a testcase for that

Comment (danhper) [reply language/move-prover/bytecode/src/access_path.rs:684]: @sblackshear I'm struggling a bit to find a test that would cover this case nicely. Let me know if you have something on the top of your head

Comment (sblackshear) [reply language/move-prover/bytecode/src/access_path.rs:684]: I think something like
```
struct S { f: address, g: address }
struct A { a: address }

fun two_secondary_indexes(s: &S, b: bool): address {
  if (b) {
    borrow_global<A>(s.f).a
  } else {
    borrow_global<A>(s.g).a
  }
}
```

will exercise this (but not positive)

File: language/tools/read-write-set/src/dynamic_analysis.rs [hunk 7]
```diff
@@ -147,8 +152,11 @@ impl ConcretizedFormals {
             } else {
                 AbsAddr::default()
             };
-            new_actuals.push(actual)
+            new_actuals.push(actual);
         }
+
+        let temp_indices = (0..formal_types.len()).collect::<Vec<usize>>();
```
Comment (sblackshear) [inline language/tools/read-write-set/src/dynamic_analysis.rs:158]: I think it would probably make more sense to do this inside `from_args_` (perhaps in the same loop that builds the `sub_map`.

File: language/move-prover/bytecode/src/read_write_set_analysis.rs [hunk 8]
```diff
@@ -134,8 +139,8 @@ impl ReadWriteSetState {
     pub fn apply_summary(
         &mut self, // caller state
         callee_summary_: &Self,
-        actuals: &[AbsAddr],
         type_actuals: &[Type],
+        temp_indices: &[TempIndex],
```
Comment (sblackshear) [inline language/move-prover/bytecode/src/read_write_set_analysis.rs:143]: Nit: could we call this `actuals` instead of `temp_indices` (here and elsewhere)? I find it slightly more readable to see that all the components of a call (type actuals, actuals, returns) are included.

File: language/move-prover/bytecode/src/read_write_set_analysis.rs [hunk 9]
```diff
@@ -81,22 +81,27 @@ impl ReadWriteSetState {
...
         // (1) bind all footprint values and types in callee accesses to their caller values
         let mut new_accesses =
-            accesses.substitute_footprint_skip_data(&actuals, type_actuals, sub_map);
+            accesses.substitute_footprint_skip_data(type_actuals, temp_indices, fun_env, sub_map);
         // (2) bind footprint paths in callee accesses with their caller values
-        for (i, actual_v) in actuals.iter().enumerate() {
-            let formal_i = Root::from_index(i, fun_env);
+        for (callee_i, caller_i) in temp_indices.iter().enumerate() {
+            let formal_i = Root::from_index(callee_i, fun_env);
+            let ap = AccessPath::from_index(*caller_i, fun_env);
```
Comment (sblackshear) [inline language/move-prover/bytecode/src/read_write_set_analysis.rs:95]:
```suggestion
            let actual_i = AccessPath::from_index(*caller_i, fun_env);
```

File: language/move-prover/bytecode/src/read_write_set_analysis.rs (hunk 8 above)
Comment (danhper) [reply language/move-prover/bytecode/src/read_write_set_analysis.rs:143]: `temp_indices` is indeed not a very useful name, I'll update that!

File: language/move-prover/bytecode/src/access_path.rs (hunk 6 above)
Comment (danhper) [reply language/move-prover/bytecode/src/access_path.rs:677]: I added the test but looking at the output I think it's not triggering the condition we discussed, the result looks correct with the previous implementation that exited the loop. I'll try to play a bit around with this.

**Comment by danhper** [comment]
@sblackshear Thanks a lot for all the comments! I think it's all resolved!

File: language/move-prover/bytecode/src/read_write_set_analysis.rs (hunk 8 above)
Comment (danhper) [reply language/move-prover/bytecode/src/read_write_set_analysis.rs]: I also reordered the arguments so that it's clear in the diff that we really just changed the type of `actuals`

File: language/move-prover/bytecode/src/access_path.rs (hunk 6 above)
Comment (sblackshear) [reply language/move-prover/bytecode/src/access_path.rs:677]: Ah, interesting... One more try;

```
struct A { b : B }
struct B { addr: address }

fun write_both_fields(a: &A, b: B, addr: address, flag: bool) {
  if (flag) {
    a.b = b
  }
  if (flag) {
    a.b.addr = addr
  }
}
```

**Comment by sblackshear** [comment]
> @sblackshear Thanks a lot for all the comments! I think it's all resolved!
> Let me know if you would prefer me to squash the commits or if multiple commits are fine, as I'm not sure which convention you are following.

Yep, LGTM! Squashing is the preferred style.

**Review by sblackshear — APPROVED** [review]

File: language/move-prover/bytecode/src/access_path.rs (hunk 6 above)
Comment (danhper) [reply language/move-prover/bytecode/src/access_path.rs:677]: I tried it but it also gives the same result with both implementations. I'm yet to find something that fails with the previous implementation but I am pretty sure there are cases where it does happen :thinking: 
I added the test to the test suite nevertheless

**Comment by danhper** [comment]
> Squashing is the preferred style.

It's now all squashed!

**Review by sblackshear — DISMISSED** [review]

**Comment by danhper** [comment]
@sblackshear The test that failed apparently timed out and seems very unrelated to this PR so I suppose it's just a matter of trying again?

**Comment by sblackshear** [comment]
Yep, I will try again (and have reported the flaky test on Slack).

/land
